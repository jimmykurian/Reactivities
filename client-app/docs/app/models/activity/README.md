[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
