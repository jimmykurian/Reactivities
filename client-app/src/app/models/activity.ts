/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
