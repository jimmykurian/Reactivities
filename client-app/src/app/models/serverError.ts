/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
