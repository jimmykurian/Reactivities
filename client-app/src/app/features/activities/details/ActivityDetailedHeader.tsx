/**
 * @file ActivityDetailedHeader.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityDetailedHeader component that displays the header section of an activity's details.
 */

import { observer } from 'mobx-react-lite';
import { Button, Header, Item, Segment, Image } from 'semantic-ui-react';
import { Activity } from '../../../models/activity';
import { Link } from 'react-router-dom';

/**
 * @constant
 * @description The styling for the activity image to apply a brightness filter.
 * @type {Object}
 */
const activityImageStyle = {
  filter: 'brightness(30%)',
};

/**
 * @constant
 * @description The styling for the text overlay on the activity image.
 * @type {Object}
 */
const activityImageTextStyle = {
  position: 'absolute',
  bottom: '5%',
  left: '5%',
  width: '100%',
  height: 'auto',
  color: 'white',
};

/**
 * @interface Props
 * @description The properties for the ActivityDetailedHeader component.
 * @property {Activity} activity - The activity object containing details to display.
 */
export interface Props {
  activity: Activity;
}

/**
 * @component ActivityDetailedHeader
 * @description The ActivityDetailedHeader component displays the header section of an activity's details, including the activity image, title, date, and host information.
 *
 * @param {Props} props - The properties object containing the activity to display.
 * @param {Activity} props.activity - The activity object containing details to display.
 *
 * @returns {JSX.Element} The rendered ActivityDetailedHeader component.
 *
 * @remarks
 * This component uses Semantic UI React components to create a visually appealing header for the activity details page. It includes an image with a darkened filter, the activity title, date, and host information. It also provides buttons for joining, canceling attendance, and managing the event.
 *
 * @example
 * Here is an example of how to use the ActivityDetailedHeader component:
 * ```tsx
 * import React from 'react';
 * import ActivityDetailedHeader from './features/activities/details/ActivityDetailedHeader';
 * import { Activity } from '../../../models/activity';
 *
 * const activity: Activity = {
 *   id: '1',
 *   title: 'Sample Activity',
 *   description: 'This is a sample activity description.',
 *   date: '2024-01-01',
 *   city: 'Sample City',
 *   venue: 'Sample Venue',
 *   category: 'Sample Category',
 * };
 *
 * const App = () => (
 *   <div>
 *     <ActivityDetailedHeader activity={activity} />
 *   </div>
 * );
 *
 * export default App;
 * ```
 */
export default observer(function ActivityDetailedHeader({
  activity,
}: Props): JSX.Element {
  return (
    <Segment.Group>
      <Segment basic attached="top" style={{ padding: '0' }}>
        <Image
          src={`/assets/categoryImages/${activity.category}.jpg`}
          fluid
          style={activityImageStyle}
        />
        <Segment style={activityImageTextStyle} basic>
          <Item.Group>
            <Item>
              <Item.Content>
                <Header
                  size="huge"
                  content={activity.title}
                  style={{ color: 'white' }}
                />
                <p>{activity.date}</p>
                <p>
                  Hosted by <strong>Bob</strong>
                </p>
              </Item.Content>
            </Item>
          </Item.Group>
        </Segment>
      </Segment>
      <Segment clearing attached="bottom">
        <Button color="teal">Join Activity</Button>
        <Button>Cancel attendance</Button>
        <Button
          as={Link}
          to={`/manage${activity.id}`}
          color="orange"
          floated="right"
        >
          Manage Event
        </Button>
      </Segment>
    </Segment.Group>
  );
});
