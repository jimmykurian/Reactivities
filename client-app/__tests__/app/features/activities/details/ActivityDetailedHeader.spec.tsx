import { render, screen } from '@testing-library/react';
import ActivityDetailedHeader, {
  Props,
} from '../../../../../src/app/features/activities/details/ActivityDetailedHeader';
import { faker } from '@faker-js/faker';
import { Activity } from '../../../../../src/app/models/activity';
import { MemoryRouter } from 'react-router-dom';

// Mocking observer from mobx-react-lite
jest.mock('mobx-react-lite', () => ({
  observer: (component: React.ComponentType) => component,
}));

describe('ActivityDetailedHeader', () => {
  // Static mock activity for consistent snapshot tests
  const staticMockActivity: Activity = {
    id: '1',
    title: 'Static Title',
    description: 'This is a static description for testing.',
    date: '2024-06-12T00:00:00.000Z',
    city: 'Static City',
    venue: 'Static Venue',
    category: 'static-category',
  };

  const generateMockActivity = (): Activity => ({
    id: faker.string.uuid(),
    title: faker.lorem.words(3),
    description: faker.lorem.sentence(),
    date: faker.date.future().toISOString(),
    city: faker.location.city(),
    venue: faker.location.streetAddress(),
    category: faker.lorem.word(),
  });

  const renderComponent = (props: Props) => {
    render(
      <MemoryRouter>
        <ActivityDetailedHeader {...props} />
      </MemoryRouter>,
    );
  };

  test('renders ActivityDetailedHeader component', () => {
    // Arrange
    const mockActivity = generateMockActivity();

    // Act
    renderComponent({ activity: mockActivity });

    // Assert
    expect(screen.getByText(mockActivity.title)).toBeInTheDocument();
    expect(screen.getByText(mockActivity.date)).toBeInTheDocument();
    expect(screen.getByText(/Hosted by/)).toBeInTheDocument();
    expect(screen.getByText(/Join Activity/i)).toBeInTheDocument();
    expect(screen.getByText(/Cancel attendance/i)).toBeInTheDocument();
    expect(screen.getByText(/Manage Event/i)).toBeInTheDocument();
  });

  test('displays the correct image for the activity category', () => {
    // Arrange
    const mockActivity = generateMockActivity();

    // Act
    renderComponent({ activity: mockActivity });

    // Assert
    const imgElement = screen.getByRole('img');
    expect(imgElement).toHaveAttribute(
      'src',
      `/assets/categoryImages/${mockActivity.category}.jpg`,
    );
  });

  test('matches snapshot', () => {
    // Arrange & Act
    const { asFragment } = render(
      <MemoryRouter>
        <ActivityDetailedHeader activity={staticMockActivity} />,
      </MemoryRouter>,
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
