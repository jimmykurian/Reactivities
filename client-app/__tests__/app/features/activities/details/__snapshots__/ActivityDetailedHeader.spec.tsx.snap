// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityDetailedHeader matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui segments"
  >
    <div
      class="ui basic top attached segment"
      style="padding: 0px;"
    >
      <img
        class="ui fluid image"
        src="/assets/categoryImages/static-category.jpg"
        style="filter: brightness(30%);"
      />
      <div
        class="ui basic segment"
        style="position: absolute; bottom: 5%; left: 5%; width: 100%; height: auto; color: white;"
      >
        <div
          class="ui items"
        >
          <div
            class="item"
          >
            <div
              class="content"
            >
              <div
                class="ui huge header"
                style="color: white;"
              >
                Static Title
              </div>
              <p>
                2024-06-12T00:00:00.000Z
              </p>
              <p>
                Hosted by 
                <strong>
                  Bob
                </strong>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui clearing bottom attached segment"
    >
      <button
        class="ui teal button"
      >
        Join Activity
      </button>
      <button
        class="ui button"
      >
        Cancel attendance
      </button>
      <a
        class="ui orange right floated button"
        href="/manage1"
        role="button"
      >
        Manage Event
      </a>
    </div>
  </div>
  ,
</DocumentFragment>
`;
